src/main/java/org/truffleruby/core/hash/library/CompactHashStore.java

```diff
@@ -0,0 +1,506 @@
+
+            insertIntoIndex(hash, kvPos);
+        }
+        this.numSlotsForIndexRebuild = (int) (this.index.length / 2 * THRESHOLD_LOAD_FACTOR_FOR_INDEX_REBUILD);
+    }
+
+    private void resizeKvStoreIfNeeded() {
+        if (kvStoreInsertionPos < kvStore.length) {
+            return;
+        }
+
+        kvStore = Arrays.copyOf(kvStore, 2 * kvStore.length);
+    }
+
+    private CompactHashStore copy() {
+        return new CompactHashStore(index.clone(), kvStore.clone(), kvStoreInsertionPos, numSlotsForIndexRebuild);
+    }
+
+    private Object deleteKvAndGetV(RubyHash hash, int hashPos, int keyKvPos) {
+        Object deletedValue = kvStore[keyKvPos + 1];
+        index[hashPos + 1] = INDEX_SLOT_DELETED_MARKER;
+        // TODO: Instead of naively nulling out the key-value, which can produce long gaps of nulls in the kvStore,
+        //       See if we can annotate each gap with its length so that iteration code can "jump" over it
+        kvStore[keyKvPos] = null;
+        kvStore[keyKvPos + 1] = null;
+
+        hash.size--;
+        return deletedValue;
+    }
+
+    /** @return the position of the first non-null key from the end of array, or a negative value if all keys are
+     *         null */
+    private int getLastKeyPos() {
+        int lastKeyPos = kvStoreInsertionPos - 2;
+        while (lastKeyPos > 0 && kvStore[lastKeyPos] == null) {
+            lastKeyPos -= 2;
+        }
+        return lastKeyPos;
+    }
+
+    /** @param h a hash code
+     * @param max the length of the index array
+     * @return An even index ranging from 0 to max-1 */
+    private static int getIndexPosFromHash(int h, int max) {
+        return h & (max - 2);
+    }
+
+    // Or -1 if key doesn't exist
+    @GenerateUncached
+    abstract static class GetKeyPosInKVStoreNode extends RubyBaseNode {
+        public abstract int execute(Object key, int hash, int hashPosInIndex, boolean compareByIdentity,
+                int[] index, Object[] kvStore);
+
+        @Specialization
+        protected int getKeyPos(
+                Object key, int hash, int hashPosInIndex, boolean compareByIdentity, int[] index, Object[] kvStore,
+                @Cached CompareHashKeysNode compareHashKeysNode) {
+            if (hashPosInIndex == -1) {
+                return -1;
+            }
+            int kvPos = index[hashPosInIndex + 1] - 1;
+            Object otherKey = kvStore[kvPos];
+
+            // identity implies equality
+            if (otherKey == key) {
+                return kvPos;
+            }
+            // but non-identity requires we test equality
+            int otherHash = index[hashPosInIndex];
+            if (compareHashKeysNode.execute(compareByIdentity, key, hash, otherKey, otherHash)) {
+                return kvPos;
+            }
+
+            // Key doesn't exist, another one whose hash collides with its hash does
+            return -1;
+        }
+    }
+
+    // Or -1 if hash doesn't exist
```

Comment:
Based on my reading of the callers, this value is the deletion marker. Please make that clear. If it's a logically different marker, please introduce a new constant for that. I'd suggest a different value from the other marker in that case to make debugging easier.